File: 946. Validate Stack Sequences/946. Validate Stack Sequences.cpp
#include <algorithm>  // std::make_heap, std::pop_heap, std::push_heap, std::sort_heap
#include <iostream>
#include <limits>
#include <list>
#include <numeric>
#include <queue>
#include <utility>  // std::pair, std::make_pair
#include <vector>

using namespace std;
using std::vector;

class Solution {
 public:
  bool validateStackSequences(vector<int>& pushed, vector<int>& popped) {
    if (pushed.size() == 0 && popped.size() == 0) {
      return true;
    }
    vector<int> myStack = {};
    int i = 0;
    int j = 0;
    while (j < popped.size()) {
      if (myStack.size() == 0) {
        myStack.push_back(pushed[i++]);
        continue;
      }
      if (myStack.back() == popped[j]) {
        myStack.pop_back();
        j++;
        if (popped.size() > 0 && i < pushed.size() && pushed[i] == popped[j]) {
          myStack.push_back(pushed[i++]);
        }
      } else {
        if (i < pushed.size()) {
          myStack.push_back(pushed[i++]);
        } else {
          break;
        }
      }
    }
    return myStack.size() == 0;
  }
};

int main(int argc, char const* argv[]) {
  vector<int> pushed = {0, 1};
  vector<int> popped = {1, 0};
  Solution mySolution;
  bool ret = mySolution.validateStackSequences(pushed, popped);
  cout << ret << endl;
  return 0;
}

File: 946. Validate Stack Sequences/946. Validate Stack Sequences.js
/**
 * @param {number[]} pushed
 * @param {number[]} popped
 * @return {boolean}
 */
var validateStackSequences = function (pushed, popped) {
    const newStack = [pushed[0]];
    let ret = true;
    let i = 1;
    while (popped.length > 0) {
        if (newStack[newStack.length - 1] === popped[0]) {
            newStack.pop();
            popped.shift();
            if (popped.length > 0 && pushed[i] === popped[0]) {
                newStack.push(pushed[i++])
            }
        } else {
            if (i < pushed.length) {
                newStack.push(pushed[i++])
            } else {
                ret = false;
                break;
            }
        }
    }
    return ret;
};

const pushed = [1, 2, 3, 4, 5]
const popped = [4, 3, 5, 1, 2]
const ret = validateStackSequences(pushed, popped)
console.log(ret)